File: frontend/src/api/logs.ts
import { api } from "./client";

/**
 * The system log (§22).
 *
 * Two things to keep in mind reading these types.
 *
 * `min_level` is **severity, not equality**: asking for `ERROR` returns errors
 * *and* criticals, because somebody who filters for errors is looking for
 * trouble and CRITICAL is more trouble. The server slices its ordered level
 * vocabulary, so the page never has to know the ranking.
 *
 * The tail is a **poll with a cursor**, not a socket. `cursor` is the id of the
 * newest line delivered; handing it back returns only what has arrived since,
 * which makes pausing honest — paused is simply not asking, and resuming asks
 * from where it stopped. A line id rather than a timestamp because two lines
 * can share a millisecond, and a timestamp cursor either repeats them or drops
 * one.
 *
 * See `services/logs.py`.
 */

export type LogLevel = "DEBUG" | "INFO" | "WARNING" | "ERROR" | "CRITICAL";

export interface LogLine {
  id: string;
  logged_at: string | null;
  level: LogLevel;
  service: string;
  logger: string | null;
  message: string;
  correlation_id: string | null;
  trace_id: string | null;
  user_id: string | null;
  host: string | null;
  environment: string;
  /** Milliseconds, as a number — a duration that sorted as text would put 9 after 1000. */
  duration_ms: number | null;
  status_code: number | null;
  /** So the table can mark which rows are worth opening without fetching either. */
  has_context: boolean;
  has_stack_trace: boolean;
}

/** One line in full. The heavy fields, fetched only when a line is opened. */
export interface LogEntry extends LogLine {
  context: Record<string, unknown>;
  stack_trace: string | null;
  span_id: string | null;
  /** The other lines from the same request — one failure is rarely one line. */
  related: LogLine[];
}

export interface LogFacetValue {
  value: string;
  count: number;
}

export interface LogsPage {
  items: LogLine[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
  sort: string;
  order: string;
  facets: Record<string, LogFacetValue[]>;
  columns: string[];
}

export interface LogCatalogue {
  fields: Array<{ name: string; label: string; kind: string }>;
  default_columns: string[];
  default_sort: string;
  /** Every level, including the ones at nought — see `LevelChips`. */
  levels: Array<{ key: LogLevel; count: number }>;
  /** What `retention.log_days` says, so the page cannot lie about its own data. */
  retention_days: number | null;
  total: number;
}

export interface LogTail {
  /** Oldest first, so a client appends without sorting. */
  items: LogLine[];
  cursor: string | null;
  /** More than one poll's worth was waiting — the viewer says so. */
  more: boolean;
}

export interface LogQuery {
  [key: string]: unknown;
  q?: string;
  min_level?: LogLevel | "";
  level?: string;
  service?: string;
  environment?: string;
  logger?: string;
  page?: number;
  page_size?: number;
  sort?: string;
  order?: string;
}

export const logsApi = {
  catalogue: (signal?: AbortSignal) =>
    api.get<LogCatalogue>("/admin/logs/catalogue", { signal }),
  list: (params: LogQuery = {}, signal?: AbortSignal) =>
    api.get<LogsPage>("/admin/logs", { params, signal }),
  /** `after` omitted on the first poll: the server returns the latest page and the cursor. */
  tail: (params: LogQuery & { after?: string } = {}, signal?: AbortSignal) =>
    api.get<LogTail>("/admin/logs/tail", { params, signal }),
  entry: (id: string, signal?: AbortSignal) =>
    api.get<LogEntry>(`/admin/logs/${id}`, { signal }),
  /**
   * Apply the retention policy now. Takes no `days`: the bound is
   * `retention.log_days`, and a parameter here would make this an arbitrary
   * delete wearing a retention policy's name.
   */
  prune: () =>
    api.post<{ removed: number; retention_days: number; kept: number }>(
      "/admin/logs/prune",
      {},
    ),
};

File: backend/src/api/logs.py
"""System log endpoints (§22).

Read-only but for one write, and that write is a *policy* being applied rather
than a row being edited: `prune` enacts `retention.log_days`, and it takes the
number of days from the setting instead of the request precisely so it cannot
become an arbitrary delete wearing a retention policy's name.

The rows here are written by `core/logsink` from the outcome of each request,
so a log line an endpoint could edit would be a log line worth nothing.
"""

from __future__ import annotations

from typing import Any

from src.core.auth import me, requires
from src.core.db import session_scope
from src.services import logs as service


@requires("logs.view")
def catalogue(app=None, operation: str = "", request=None, **_: Any):
    """The filter vocabulary, the per-level counts and the retention bound."""
    with session_scope() as session:
        return service.catalogue(session, principal=me()), 200


@requires("logs.view")
def collection(app=None, operation: str = "", request=None, **_: Any):
    """One page of the log: filtered, faceted, sorted and paged in PostgreSQL."""
    args = request.args if request is not None else {}
    with session_scope() as session:
        return service.listing(session, args, principal=me()), 200


@requires("logs.view")
def tail(app=None, operation: str = "", request=None, **_: Any):
    """Whatever has arrived since the caller's cursor — the live tail (§22).

    A poll rather than a socket, so pausing is simply not asking and resuming
    asks from where it stopped. See the service docstring for why.
    """
    args = request.args if request is not None else {}
    with session_scope() as session:
        return service.tail(session, args, principal=me()), 200


@requires("logs.view")
def item(app=None, operation: str = "", request=None, line_id: str = "", **kwargs: Any):
    """One line in full: its context, its stack trace and its siblings."""
    identifier = line_id or str(kwargs.get("line_id") or "")
    with session_scope() as session:
        return service.entry(session, identifier, principal=me()), 200


@requires("logs.view", "settings.manage")
def prune(app=None, operation: str = "", request=None, **_: Any):
    """Apply the retention policy now (§22).

    Two permissions: reading the log and enacting the bound on it are different
    privileges, and the bound is a setting — somebody who may not change
    `retention.log_days` should not be able to apply it early either.
    """
    args = request.args if request is not None else {}
    with session_scope() as session:
        return service.prune(session, args, principal=me()), 200
